### PR TITLE
Ensure non-fatal exceptions are actually thrown

### PR DESCRIPTION
**Pull request by tmeschter**
**Customer scenario**

No particular customer scenario, but this improves data collection in a number of cases where we might be hitting non-fatal Watsons.

When a process crashes due to an unhandled exception or an explicit `Environment.FailFast()` or similar, the collected call stack and (possible) memory dump will be very close to the point of failure. This is obviously desirable when you need to diagnose an issue from crash reports.

When we catch an exception and report it as a non-fatal event, however, that collection may happen well after the failure. There are a couple of things we need to do to ensure we get the proper data.

First, we need to ensure that the exceptions we report have actually been thrown. The call stack within an exception is populated _when it is thrown_, not when it is created. As the analysis tools on the back end use the exception stack trace to bucketize failures, we currently have a large number of reports where we know something went wrong, but we don't know where.

Second, our reporting mechanisms are best called from within an exception filter. This is because the filter runs at the point where the exception was thrown, before control flow has passed to a catch block. If we collect a memory dump at this point it will preserve the stack frames that led to the exception which is very useful for debugging.

The resulting pattern is often a little odd: we throw an exception, report it through the exception filter, immediately catch it, and then do nothing in the catch block. However, this ensures we collect actionable data.

**Bugs this fixes:**

N/A

**Workarounds, if any**

N/A

**Risk**

Low; this code only comes into play when something has already gone wrong and we're already trying to collect data for a non-fatal Watson.

**Performance impact**

Low; this code only comes into play when something has already gone wrong and we're already trying to collect data for a non-fatal Watson.

**Is this a regression from a previous update?**

N/A

**Root cause analysis:**

It's not immediately obvious how to structure code to capture data properly in a non-fatal Watson. It is easy to forget than an exception needs to be thrown before it will have a stack.

**How was the bug found?**

Looking at telemetry events I saw some cases where we weren't getting a call stack. The most likely explanation was that we were reporting an exception that had never been thrown.

### CODE REVIEW

**Comment by tmeschter** [comment]
@dotnet/roslyn-ide Please review.

File: src/EditorFeatures/Core/Implementation/Preview/DifferenceViewerPreview.cs [hunk 1]
```diff
@@ -41,7 +41,13 @@ public void Dispose()
                 return;
             }
 
-            FatalError.ReportWithoutCrash(new Exception($"Dispose is not called how? viewer state : {_viewer.IsClosed}"));
+            try
+            {
+                throw new Exception($"Dispose is not called how? viewer state : {_viewer.IsClosed}");
+            }
+            catch (Exception e) when (FatalError.ReportWithoutCrash(e))
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/Preview/DifferenceViewerPreview.cs]: :bulb: I believe the `(Exception e)` here is unnecessary.

File: src/EditorFeatures/Core/SymbolSearch/SymbolSearchUpdateEngine.Update.cs [hunk 2]
```diff
@@ -587,8 +586,13 @@ private void ParsePatchElement(XElement patchElement, out bool upToDate, out boo
                 var contentsAttribute = element.Attribute(ContentAttributeName);
                 if (contentsAttribute == null)
                 {
-                    _service._reportAndSwallowException(
```
Comment (sharwell) [inline src/EditorFeatures/Core/SymbolSearch/SymbolSearchUpdateEngine.Update.cs]: ❓ Would it break bucketing to just modify the implementation of `_reportAndSwallowException`?

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs [hunk 3]
```diff
@@ -248,7 +248,13 @@ private void OnConnectionChanged(object sender, bool connected)
 
                     // s_lastRemoteClientTask info should be saved in the dump
                     // report NFW when connection is closed unless it is proper shutdown
-                    FatalError.ReportWithoutCrash(new Exception("Connection to remote host closed"));
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: ❓ Would it break bucketing to modify the implementation of `ReportWithoutCrash`?

File: src/EditorFeatures/Core/SymbolSearch/SymbolSearchUpdateEngine.Update.cs (hunk 2 above)
Comment (tmeschter) [inline src/EditorFeatures/Core/SymbolSearch/SymbolSearchUpdateEngine.Update.cs]: @sharwell In practice `_reportAndSwallowException` is actually `FatalError.ReportWithoutCrash` so I'll address my response to your other question.

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs (hunk 3 above)
Comment (tmeschter) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: @sharwell That would depend on the modification. Bucketing depends on a system of heuristics that we largely do not control, though process name, assembly name, exception name, and call stack figure prominently. If we leave those alone bucketing probably stays the same; if we change them, reports will almost certainly start flowing into other buckets.

At the moment changing the bucket isn't really a problem, though, as the places I've fixed up are basically producing no useful data. This change is really just about making all the places we try to report a non-fatal error actually work on the most basic level. :-)

Did I answer your question, or were you referring to something else?

File: src/EditorFeatures/Core/Implementation/Preview/DifferenceViewerPreview.cs (hunk 1 above)
Comment (tmeschter) [inline src/EditorFeatures/Core/Implementation/Preview/DifferenceViewerPreview.cs]: No, I need `e` so that it can be passed to `FatalError.ReportWithoutCrash`.

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs (hunk 3 above)
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: It seems like your change would be *almost* equivalent to just adding the following to the beginning of `FatalError.ReportWithoutCrash`:

```csharp
if (exception.StackTrace == null)
{
    try
    {
        throw exception;
    }
    catch
    {
        // Empty... we just needed the exception to have a stack trace
    }
}
```

The only difference between what you wrote and this would be the inclusion of `ReportWithoutCrash` at the top of the stack trace. If that would still work, it definitely makes the code easier to read.

Comment (tmeschter) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: Ah, I see.

I could see adding a new helper method on `FatalError` to simplify this. `FatalError.ReportWithoutCrash` was really always meant to be called in an exception filter; I don't really want to overload its meaning.

There are other changes I'm mulling over for how we report non-fatal exceptions, and such a helper would fit in better with those. So I'll keep it in mind.

**Comment by tmeschter** [comment]
@dotnet-bot retest please

**Comment by tmeschter** [comment]
@dotnet-bot test please

**Comment by tmeschter** [comment]
@dotnet-bot test this please

**Comment by tmeschter** [comment]
@dotnet/roslyn-ide Please review.

**Comment by tmeschter** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by jasonmalinowski** [comment]
I think I echo @sharwell's thoughts here: if there isn't a stack trace, it seems we probably should throw and recatch inside ReportWithoutCrash. Otherwise, there's zero protection of this mistake being made again.

**Comment by tmeschter** [comment]
@jasonmalinowski @sharwell I'll make that change.

**Comment by tmeschter** [comment]
@jasonmalinowski @sharwell Do you think this is good as-is, or are you interested in seeing further changes?

File: src/EditorFeatures/Core/Implementation/Preview/DifferenceViewerPreview.cs [hunk 4]
```diff
@@ -41,7 +41,13 @@ public void Dispose()
                 return;
             }
 
-            FatalError.ReportWithoutCrash(new Exception($"Dispose is not called how? viewer state : {_viewer.IsClosed}"));
+            try
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/Preview/DifferenceViewerPreview.cs]: We can undo these changes now, right?

File: src/EditorFeatures/Core/SymbolSearch/SymbolSearchUpdateEngine.Update.cs [hunk 5]
```diff
@@ -564,15 +564,14 @@ private void ParsePatchElement(XElement patchElement, out bool upToDate, out boo
...
+                    catch (Exception e) when (IOUtilities.IsNormalIOException(e) || _service._reportAndSwallowException(e))
                     {
-                        // Normal IO exception. Don't bother reporting it.  We don't want to get
-                        // lots of hits just because we couldn't write a file because of something
-                        // like an anti-virus tool lockign the file.
-                        if (!IOUtilities.IsNormalIOException(e))
-                        {
-                            _service._reportAndSwallowException(e);
-                        }
+                        // The exception filter above might be a little funny looking. We always
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/SymbolSearch/SymbolSearchUpdateEngine.Update.cs]: "might" is being generous here... :-)

File: src/EditorFeatures/Core/Implementation/Preview/DifferenceViewerPreview.cs (hunk 4 above)
Comment (tmeschter) [inline src/EditorFeatures/Core/Implementation/Preview/DifferenceViewerPreview.cs]: @jasonmalinowski I've gone back and removed the unnecessary try/catches.

**Comment by tmeschter** [comment]
@jasonmalinowski I've gone back and removed the unnecessary try/catch blocks.

**Comment by tmeschter** [comment]
Tagging @MattGertz for approval.

**Comment by MattGertz** [comment]
Approved pending tests.

File: src/Compilers/Core/Portable/InternalUtilities/FatalError.cs [hunk 6]
```diff
@@ -131,13 +131,34 @@ public static bool Report(Exception exception)
...
         /// </summary>
         /// <returns>True to catch the exception.</returns>
         [DebuggerHidden]
         public static bool ReportWithoutCrash(Exception exception)
         {
+            // There have been cases where a new, unthrown exception has been passed to this method.
+            // In these cases the exception won't have a stack trace, which isn't very helpful. We
+            // throw and catch the exception here as that will result in a stack trace that is
+            // better than nothing.
+            if (exception.StackTrace == null)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/FatalError.cs]: ❔ Should this be added to the other reporting methods?

❔ Can we add a regression test for this behavior, since a regression is likely to not be visible prior to shipping?

File: src/EditorFeatures/Core/SymbolSearch/SymbolSearchUpdateEngine.Update.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/SymbolSearch/SymbolSearchUpdateEngine.Update.cs]: BTW, i'm all for doing this in a better way.  The point of this was that it allowed me to actually stringently test the error paths of SymbolSearch from tests.  i.e. tests could simulate different sorts of exceptions happening at different times, and could verify that hte engine would do the right thing.

**Comment by tmeschter** [comment]
@dotnet-bot test windows_eta_open_prtest please

**Comment by tmeschter** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please
